aten/src/ATen/copy_wrapper.py

```diff
@@ -19,90 +16,199 @@
+                # redispatch
+                continue
+
+        body_env = nested_dict({
+            'src_scalar_name': src_type['ScalarName'],
+            'case_id': src_type['TypeID'],
+            'src_tensor': src_type['Tensor'],
+            'dst_tensor': dst_type['Tensor'],
+            'cuda': cuda,
+            'state': state,
+        }, dst_type)
+
+        copies = []
+        if dst_type['ScalarType'] == src_type['ScalarType']:
+            if dst_type['Backend'] == 'CUDA' and src_type['Backend'] == 'CPU':
+                copies.append(COPY_ASYNC_CPU.substitute(body_env))
+        copies.append(COPY.substitute(body_env))
+
+        copy_body.append(CASE.substitute(body_env, copies=copies))
+
+    if dst_type['Backend'] == 'CPU':
+        # CPU fallthrough needs to redispatch to s_copy_from
+        # (Backend == CPU implies Dense)
+        assert dst_type['Density'] == 'Dense'
+        function_fallthrough = FUNCTION_FALLTHROUGH_REDISPATCH
+    else:
+        function_fallthrough = FUNCTION_FALLTHROUGH_ERROR
+
+    env = nested_dict({
+        'function_fallthrough': function_fallthrough
+    }, dst_type)
+    return FUNCTION.substitute(env, copy_body=copy_body)
+
+
+def create_one_copy_from(src_type, all_types):
+    if src_type['DenseBackend'] == 'CPU':
+        return FUNCTION_FROM_SWAP.substitute(src_type)
+
+    copy_body = []
+
+    for dst_type in all_types:
+        if dst_type['Density'] == 'Sparse' or src_type['Density'] == 'Sparse':
+            # skip sparse copies, which are not yet implemented
+            continue
         cuda = ''
+        state = []
         if src_type['Backend'] == 'CUDA':
             cuda = 'Cuda'
-        if env['Backend'] == 'CUDA' or src_type['Backend'] == 'CUDA':
-            state.append('context->thc_state')
+        if dst_type['Backend'] == 'CUDA' or src_type['Backend'] == 'CUDA':
+            state = ['context->thc_state']
 
-        combined = nested_dict({
+        body_env = nested_dict({
             'src_scalar_name': src_type['ScalarName'],
-            'src_id': src_type['TypeID'],
+            'case_id': dst_type['TypeID'],
             'src_tensor': src_type['Tensor'],
+            'dst_tensor': dst_type['Tensor'],
             'cuda': cuda,
             'state': state,
-        }, env)
+        }, dst_type)
 
         copies = []
-        if env['ScalarType'] == src_type['ScalarType']:
-            if env['Backend'] == 'CUDA' and src_type['Backend'] == 'CPU':
-                copies.append(COPY_ASYNC_CPU.substitute(combined))
-            if env['Backend'] == 'CPU' and src_type['Backend'] == 'CUDA':
-                copies.append(COPY_ASYNC_CUDA.substitute(combined))
-        copies.append(COPY.substitute(combined))
-
-        copy_body.append(CASE.substitute(combined, copies=copies))
-    return FUNCTION.substitute(env, copy_body=copy_body)
+        if dst_type['ScalarType'] == src_type['ScalarType']:
+            if dst_type['Backend'] == 'CPU':
```

Comment:
nit: aren't you providing CPU to CPU here in this function (okay, it's early exited, but whatever); seems safter to do the and src_type['Backend'] == 'CUDA' thing here in case the structure changes.

Reply:
Sure, I can do that (though I am hoping this entire Python file dies before we add a new backend :)

Reply:
By that logic, you should just check in the generated code :).

Reply:
I seriously considered it. Are you suggesting we do it? :)

Reply:
If you feel strongly about it, do it.  Otherwise, I'd just leave this as is.